cmd/webhook/admission/admission.go

```diff
@@ -0,0 +1,220 @@
+		Name:      "requests",
+		Help:      "Total number of requests to this admitter",
+	}, []string{"admitter"})
+	invalidRequests = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: MetricNamespace,
+		Subsystem: metricSubsystem,
+		Name:      "invalid_requests",
+		Help:      "Total number of requests to this admitter that couldn't be parsed",
+	}, []string{"admitter"})
+	rejectedRequests = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: MetricNamespace,
+		Subsystem: metricSubsystem,
+		Name:      "rejected_admissions",
+		Help:      "Total number of requests rejected by this admitter",
+	}, labels)
+	approvedRequests = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: MetricNamespace,
+		Subsystem: metricSubsystem,
+		Name:      "successful_admissions",
+		Help:      "Total number of requests successfully processed by this admitter",
+	}, labels)
+	admissionDuration = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: MetricNamespace,
+		Subsystem: metricSubsystem,
+		Name:      "admission_duration",
+		Help:      "Duration of admission calls",
+		Buckets:   []float64{.005, .01, .025, .05, .1, .25, .5, .75, 1, 1.25, 1.5, 2, 2.5, 5, 10},
+	}, labels)
+)
+
+type Admitter interface {
+	Name() string
+	Admit(req *admissionsv1.AdmissionRequest) (*admissionsv1.AdmissionResponse, error)
+}
+
+type RouteGroupAdmitter struct {
+}
+
+func init() {
+	prometheus.MustRegister(totalRequests, invalidRequests, rejectedRequests, approvedRequests, admissionDuration)
+}
+
+func (r RouteGroupAdmitter) Name() string {
+	return "routegroup"
+}
+
+func (r RouteGroupAdmitter) Admit(req *admissionsv1.AdmissionRequest) (*admissionsv1.AdmissionResponse, error) {
+	rgItem := definitions.RouteGroupItem{}
+	err := json.Unmarshal(req.Object.Raw, &rgItem)
+	if err != nil {
+		emsg := fmt.Sprintf("could not parse RouteGroup, %v", err)
+		log.Error(emsg)
+		return &admissionsv1.AdmissionResponse{
+			UID:     req.UID,
+			Allowed: false,
+			Result: &metav1.Status{
+				Message: emsg,
+			},
+		}, nil
+	}
+
+	err = definitions.ValidateRouteGroup(&rgItem)
+	if err != nil {
+		emsg := fmt.Sprintf("could not validate RouteGroup, %v", err)
+		log.Error(emsg)
+		return &admissionsv1.AdmissionResponse{
+			UID:     req.UID,
+			Allowed: false,
+			Result: &metav1.Status{
+				Message: emsg,
+			},
+		}, nil
+	}
+
+	return &admissionsv1.AdmissionResponse{
+		UID:     req.UID,
+		Allowed: true,
+	}, nil
+}
+
+func Handler(admitter Admitter) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		admitterName := admitter.Name()
+		totalRequests.WithLabelValues(admitterName).Inc()
+
+		if r.Method != "POST" || r.Header.Get("Content-Type") != "application/json" {
+			invalidRequests.WithLabelValues(admitterName).Inc()
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			log.Errorf("Unable to read request: %v", err)
```

Comment:
Normally we use `Failed to ...` instead of `Unable to...`, I think this should be fixed.